### ISSUE TITLE
question about "possibility" and "min_possibility" used in file "main_SemanticKITTI.py"

### ISSUE THREAD

**Issue by amiltonwong** [QingyongHu/RandLA-Net#11]
Hi, @QingyongHu ,

Could you help to answer the following questions about "possibility" and "min_possibility' used in file "main_SemanticKITTI.py"?

Q1. What's the usage for [variable](https://github.com/QingyongHu/RandLA-Net/blob/master/main_SemanticKITTI.py#L35-L36) "possibilities" and "min_possibility"? 

Q2. They are found in [data loading](https://github.com/QingyongHu/RandLA-Net/blob/master/main_SemanticKITTI.py#L51-L52) in test set. What does it help?

Q3. They are also found in [spatially_regular_gen()](https://github.com/QingyongHu/RandLA-Net/blob/master/main_SemanticKITTI.py#L65-L75). Could you elaborate their functions?

THX!

**Comment by QingyongHu** [comment]
Hi @amiltonwong. Thanks for your interest in our work. The main reason we use this function [spatially_regular_gen](https://github.com/QingyongHu/RandLA-Net/blob/04829e8bb57f802ff406a2f0aff4a5f64ef65179/main_SemanticKITTI.py#L65-L75) is to make sure that all the points could be tested efficiently, otherwise, it will take a long time to test all the points if we randomly crop a small point cloud for each scan at a time.

Specifically, we first randomly generate a value for each point [(Line 51)](https://github.com/QingyongHu/RandLA-Net/blob/04829e8bb57f802ff406a2f0aff4a5f64ef65179/main_SemanticKITTI.py#L51) which represents the probability of this point being selected. Then, we choose the point with the lowest probability as the center point each time during test [(Line 66)](https://github.com/QingyongHu/RandLA-Net/blob/04829e8bb57f802ff406a2f0aff4a5f64ef65179/main_SemanticKITTI.py#L66). Next, a fixed number of points are selected based on KDTree [(Line 69)](https://github.com/QingyongHu/RandLA-Net/blob/04829e8bb57f802ff406a2f0aff4a5f64ef65179/main_SemanticKITTI.py#L69), we then increase the probability of the queried point, this value is calculated based on the distance of the center point [(Line71-75)](https://github.com/QingyongHu/RandLA-Net/blob/04829e8bb57f802ff406a2f0aff4a5f64ef65179/main_SemanticKITTI.py#L71). Then we continue to select the point with the lowest probability as the center point (In this case, the points already selected are unlikely to be selected again), this procedure is iterated to make sure all points were tested.

I hope this would be helpful.

Many thanks for your interest.

**Comment by amiltonwong** [comment]
@QingyongHu , Thanks a lot for your detailed explanation :)

**Comment by jingyibo123** [comment]
Hi @QingyongHu , In the code this mecanism is only used for testing. 
The way I understand it is during training, with enough epoch random cropping should eventually cover all points.
However since the model are agnostic to the number of points, it should be theoretically possible to infer all points of a scan without any cropping at all, right?

**Comment by QingyongHu** [comment]
Hi @jingyibo123, 
Yes! You are free to change the `num_points `during inference, or just inference the whole point clouds if you have enough GPU resources

**Comment by Haiyan-Chris-Wang** [comment]
Hi @QingyongHu, thanks for your work. I have quite a small question here regard to this discussion, would the performance be different if you choose the above-mentioned min_possibility trick or just directly inference the whole point clouds? I am thinking that maybe the min_possibility trick can help your testing data have the same organization and size with your training data. Thus the performance is also benefited from this trick.
